crypto/evp/pkey_kdf.c

```diff
@@ -11,131 +11,249 @@
     EVP_KDF_CTX *kctx;
+    const char *kdf_name = OBJ_nid2sn(ctx->pmeth->pkey_id);
+    EVP_KDF *kdf;
 
-    kctx = EVP_KDF_CTX_new_id(ctx->pmeth->pkey_id);
-    if (kctx == NULL)
+    pkctx = OPENSSL_zalloc(sizeof(*pkctx));
+    if (pkctx == NULL)
         return 0;
 
-    ctx->data = kctx;
+    kdf = EVP_KDF_fetch(NULL, kdf_name, NULL);
+    kctx = EVP_KDF_CTX_new(kdf);
+    EVP_KDF_free(kdf);
+    if (kctx == NULL) {
+        OPENSSL_free(pkctx);
+        return 0;
+    }
+
+    pkctx->kctx = kctx;
+    ctx->data = pkctx;
     return 1;
 }
 
 static void pkey_kdf_cleanup(EVP_PKEY_CTX *ctx)
 {
-    EVP_KDF_CTX *kctx = ctx->data;
+    EVP_PKEY_KDF_CTX *pkctx = ctx->data;
 
-    EVP_KDF_CTX_free(kctx);
+    EVP_KDF_CTX_free(pkctx->kctx);
+    pkey_kdf_free_collected(pkctx);
+    OPENSSL_free(pkctx);
+}
+
+static int collect(BUF_MEM **collector, void *data, size_t datalen)
+{
+    size_t i;
+
+    if (*collector == NULL)
+        *collector = BUF_MEM_new();
+    if (*collector == NULL) {
+        ERR_raise(ERR_LIB_EVP, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+
+    i = (*collector)->length; /* BUF_MEM_grow() changes it! */
+    /*
+     * The i + datalen check is to distinguish between BUF_MEM_grow()
+     * signaling an error and BUF_MEM_grow() simply returning the (zero)
+     * length.
+     */
+    if (!BUF_MEM_grow(*collector, i + datalen)
+        && i + datalen != 0)
+        return 0;
+    if (data != NULL)
+        memcpy((*collector)->data + i, data, datalen);
+    return 1;
 }
 
 static int pkey_kdf_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
 {
-    EVP_KDF_CTX *kctx = ctx->data;
-    uint64_t u64_value;
-    int cmd;
-    int ret;
+    EVP_PKEY_KDF_CTX *pkctx = ctx->data;
+    EVP_KDF_CTX *kctx = pkctx->kctx;
+    enum { T_OCTET_STRING, T_UINT64, T_DIGEST, T_INT } cmd;
+    const char *name, *mdname;
+    BUF_MEM **collector = NULL;
+    OSSL_PARAM params[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
 
     switch (type) {
     case EVP_PKEY_CTRL_PASS:
-        cmd = EVP_KDF_CTRL_SET_PASS;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_PASSWORD;
         break;
     case EVP_PKEY_CTRL_HKDF_SALT:
     case EVP_PKEY_CTRL_SCRYPT_SALT:
-        cmd = EVP_KDF_CTRL_SET_SALT;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_SALT;
         break;
     case EVP_PKEY_CTRL_TLS_MD:
     case EVP_PKEY_CTRL_HKDF_MD:
-        cmd = EVP_KDF_CTRL_SET_MD;
+        cmd = T_DIGEST;
+        name = OSSL_KDF_PARAM_DIGEST;
         break;
     case EVP_PKEY_CTRL_TLS_SECRET:
-        cmd = EVP_KDF_CTRL_SET_TLS_SECRET;
-        ret = EVP_KDF_ctrl(kctx, EVP_KDF_CTRL_RESET_TLS_SEED);
-        if (ret < 1)
-            return ret;
+        cmd = T_OCTET_STRING;
+        name = OSSL_KDF_PARAM_SECRET;
+        /*
+         * Perform the semantics described in
+         * EVP_PKEY_CTX_add1_tls1_prf_seed(3)
+         */
+        if (ctx->pmeth->pkey_id == NID_tls1_prf)
+            BUF_MEM_free(pkctx->collected_seed);
```

Comment:
Oh, I see that I forgot to do `pkctx->collected_seed = NULL`... Please add such a line

Reply:
Okay.